Add announcement blog post for Flink 1.19 (#721)
apache/flink-web: docs/content/posts/2024-03-xx-release-1.19.0.md

> ---

Reviewer: @lincoln-lil I think we're missing something on the changes to the Sink API, from both https://issues.apache.org/jira/browse/FLINK-25857 and https://issues.apache.org/jira/browse/FLINK-33972. Same for https://cwiki.apache.org/confluence/display/FLINK/FLIP-400%3A+AsyncScalarFunction+for+asynchronous+scalar+function+support.

LGTM otherwise

Author: @MartijnVisser Thanks for reviewing this! Yes, these changes need to be added.
For FLIP-400, I've added a paragraph into the sql section based on the flip page.
For the two changes related to the sink api(FLINK-25857 & FLINK-33973), do we put both in the deprecation section?

Author: @MartijnVisser As we discussed at the sync meeting, I've added these descriptions (with the minor tweak of using 'Connector API Improvements' as the title, rather than the broader 'SDK Improvements').

authors:
- LincolnLee:
  name: "Lincoln Lee"
  twitter: lincoln_86xy

date: "2024-03-xxT22:00:00Z"
subtitle: ""
title: Announcing the Release of Apache Flink 1.19
aliases:
- /news/2024/03/xx/release-1.19.0.html
---

The Apache Flink PMC is pleased to announce the release of Apache Flink 1.19.0. As usual, we are
looking at a packed release with a wide variety of improvements and new features. Overall, 162

> people contributed to this release completing 33 FLIPS and 600+ issues. Thank you!

Reviewer 2: ```suggestion
people contributed to this release completing 33 FLIPs and 600+ issues. Thank you!
```

Let's dive into the highlights.

# Flink SQL Improvements

## Custom Parallelism for Table/SQL Sources

Now in Flink 1.19, you can set a custom parallelism for performance tuning via the `scan.parallelism`
option. The first available connector is DataGen (Kafka connector is on the way). Here is an example
using SQL Client:

```sql
-- set parallelism within the ddl
CREATE TABLE Orders (
    order_number BIGINT,
    price        DECIMAL(32,2),
    buyer        ROW<first_name STRING, last_name STRING>,
    order_time   TIMESTAMP(3)
) WITH (
    'connector' = 'datagen',
    'scan.parallelism' = '4'
);

-- or set parallelism via dynamic table option

> SELECT * FROM Orders /*+ OPTIONS('scan.parallelism'='4') */;

Reviewer 2: I'd suggest to keep the announcement concise and move all the sample codes to detailed documentations.

Author: I've checked the last few release announcements and we've included some simple examples in some of the feature introductions since 1.18(https://flink.apache.org/2023/10/24/announcing-the-release-of-apache-flink-1.18/), which from a user's point of view is probably more of an introduction to what a feature looks like, rather than a complete description of how to use the feature.
I'm not sure how to choose between simplicity and readability, maybe we need to hear more feedback on this, or discuss it in the next release sync and I'll change it accordingly : )

```

**More Information**
* [Documentation](https://nightlies.apache.org/flink/flink-docs-release-1.19/docs/dev/table/sourcessinks/#scan-table-source)
* [FLIP-367: Support Setting Parallelism for Table/SQL Sources](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=263429150)

## Configurable SQL Gateway Java Options

> A new option `env.java.opts.sql-gateway` for specifying the Java options, so you can fine-tune the

Reviewer 2: ```suggestion
A new option `env.java.opts.sql-gateway` for specifying the Java options is introduced in Flink 1.19, so you can fine-tune the
```

* [FLINK-33203](https://issues.apache.org/jira/browse/FLINK-33203)

## Configure Different State TTLs using SQL Hint

Starting from Flink 1.18, Table API and SQL users can set state time-to-live (TTL) individually for
stateful operators via the SQL compiled plan. In Flink 1.19, users have a more flexible way to
specify custom TTL values for regular joins and group aggregations directly within their queries by [utilizing the STATE_TTL hint](https://nightlies.apache.org/flink/flink-docs-release-1.19/docs/dev/table/sql/queries/hints/#state-ttl-hints).
This improvement means that you no longer need to alter your compiled plan to set specific TTLs for
these frequently used operators. With the introduction of `STATE_TTL` hints, you can streamline your workflow and
dynamically adjust the TTL based on your operational requirements.

Here is an example:
```sql
-- set state ttl for join
SELECT /*+ STATE_TTL('Orders'= '1d', 'Customers' = '20d') */ *
FROM Orders LEFT OUTER JOIN Customers
    ON Orders.o_custkey = Customers.c_custkey;

-- set state ttl for aggregation
SELECT /*+ STATE_TTL('o' = '1d') */ o_orderkey, SUM(o_totalprice) AS revenue
FROM Orders AS o
GROUP BY o_orderkey;

> ```

Reviewer 2: Same here for the sample code.

## Named Parameters for Functions and Procedures

Named parameters now can be used when calling a function or stored procedure. With named parameters,
users do not need to strictly specify the parameter position, just specify the parameter name and its
corresponding value. At the same time, if non-essential parameters are not specified, they will default to being filled with null.

Here's an example of defining a function with one mandatory parameter and two optional parameters using named parameters:
```java
public static class NamedArgumentsTableFunction extends TableFunction<Object> {

	@FunctionHint(
			output = @DataTypeHint("STRING"),
			arguments = {
					@ArgumentHint(name = "in1", isOptional = false, type = @DataTypeHint("STRING")),
					@ArgumentHint(name = "in2", isOptional = true, type = @DataTypeHint("STRING")),
					@ArgumentHint(name = "in3", isOptional = true, type = @DataTypeHint("STRING"))})
	public void eval(String arg1, String arg2, String arg3) {
		collect(arg1 + ", " + arg2 + "," + arg3);
	}

}

> ```

Reviewer 2: Same here for the sample code.

When calling the function in SQL, parameters can be specified by name, for example:
```sql
SELECT * FROM TABLE(myFunction(in1 => 'v1', in3 => 'v3', in2 => 'v2'))
```
Also the optional parameters can be omitted:
```sql
SELECT * FROM TABLE(myFunction(in1 => 'v1'))
```

**More Information**
* [Documentation](https://nightlies.apache.org/flink/flink-docs-release-1.19/docs/dev/table/functions/udfs/#named-parameters)
* [FLIP-387: Support named parameters for functions and call procedures](https://cwiki.apache.org/confluence/display/FLINK/FLIP-387%3A+Support+named+parameters+for+functions+and+call+procedures)

## Window TVF Aggregation Features

* Supports SESSION Window TVF in Streaming Mode

> Now users can use SESSION Window TVF in streaming mode. A simple example is as follows:

Reviewer 2: I don't see this blogpost when trying to build the website, which might relates to the date issues. However, according to my experience, this line break in the markdown file won't lead to a line break in a built web-page, which could be confusing to the readers.

Author: You're right, a line break is needed here and also following positions(this blogpost located on the last page due to the unsettled date).

```sql
-- session window with partition keys
SELECT * FROM TABLE(
   SESSION(TABLE Bid PARTITION BY item, DESCRIPTOR(bidtime), INTERVAL '5' MINUTES));

-- apply aggregation on the session windowed table with partition keys
SELECT window_start, window_end, item, SUM(price) AS total_price
FROM TABLE(
    SESSION(TABLE Bid PARTITION BY item, DESCRIPTOR(bidtime), INTERVAL '5' MINUTES))
GROUP BY item, window_start, window_end;

> ```

Reviewer 2: Same here for the sample code.

* Supports Changelog Inputs For Window TVF Aggregation

>   Window aggregation operators (generated based on Window TVF Function) can now handle changelog

Reviewer 2: Same here for the line break.

>   streams (e.g., CDC data sources, etc.) without issue. Users are recommended to migrate from legacy

Reviewer 2: ```suggestion
  streams (e.g., CDC data sources, etc.). Users are recommended to migrate from legacy
```

**More Information**
* [flink-configuration-file](https://nightlies.apache.org/flink/flink-docs-release-1.19/docs/deployment/config/#flink-configuration-file)
* [FLIP-366: Support standard YAML for FLINK configuration](https://cwiki.apache.org/confluence/display/FLINK/FLIP-366%3A+Support+standard+YAML+for+FLINK+configuration?src=contextnavpagetreemode)

## Profiling JobManager/TaskManager on Flink Web

In Flink 1.19, we support triggering profiling at the JobManager/TaskManager level, allowing users to
create a profiling instance with arbitrary intervals and event modes (supported by [async-profiler](https://github.com/async-profiler/async-profiler)).
Users can easily submit profiles and export results in the Flink Web UI.

For example,
- First, users should identify the candidate TaskManager/JobManager with performance bottleneck for profiling and switch to the corresponding TaskManager/JobManager page (profiler tab).
- The user simply clicks on the `Create Profiling Instance` button to submit a profiling instance with specified period and mode. (The description of the profiling mode will be displayed when hovering over the corresponding mode.)

> - Once the profiling instance is complete, the user can easily download the interactive HTML file by clicking on the link.

Reviewer 2: Again, for the example. IMHO, this announcement should focus on telling users what's new in this release, rather than teaching users how to use these new features.

<div style="text-align: center;">
<img src="/img/blog/2024-03-xx-release-1.19.0/profiling.png" style="width:90%;margin:15px">
</div>

Profiling result:
<div style="text-align: center;">
<img src="/img/blog/2024-03-xx-release-1.19.0/profiling-res.png" style="width:90%;margin:15px">
</div>

**More Information**
* [Documentation](https://nightlies.apache.org/flink/flink-docs-release-1.19/docs/ops/debugging/profiler/)
* [FLIP-375: Built-in cross-platform powerful java profiler](https://cwiki.apache.org/confluence/x/64lEE)

## New Config Options for Administrator JVM Options

A set of administrator JVM options are available, which prepend the user-set extra JVM options for
platform-wide JVM tuning.

**More Information**
* [Documentation](https://nightlies.apache.org/flink/flink-docs-release-1.19/docs/deployment/config/#jvm-and-logging-options)
* [FLIP-397: Add config options for administrator JVM options](https://cwiki.apache.org/confluence/display/FLINK/FLIP-397%3A+Add+config+options+for+administrator+JVM+options?src=jira)

# Checkpoints Improvements

> ## Using Larger Checkpointing Interval When Source is Processing Backlog
> 
> `ProcessingBacklog` is introduced to demonstrate whether a record should be processed with low latency
> or high throughput. `ProcessingBacklog` can be set by source operators, and can be used to change the
> checkpoint interval of a job during runtime.
> 
> **More Information**
> * [FLINK-32514](https://issues.apache.org/jira/browse/FLINK-32514)
> * [FLIP-309: Support using larger checkpointing interval when source is processing backlog](https://cwiki.apache.org/confluence/display/FLINK/FLIP-309%3A+Support+using+larger+checkpointing+interval+when+source+is+processing+backlog

Reviewer 3: ```suggestion
## Using Larger Checkpointing Interval When Source is Processing Backlog

`IsProcessingBacklog` is introduced to indicate whether a record should be processed with low latency 
or high throughput. Connector developers can update the Source implementation to utilize the 
`SplitEnumeratorContext#setIsProcessingBacklog` method to report whether the records are backlog records. 
Users can set the `execution.checkpointing.interval-during-backlog` to use a larger checkpoint interval 
to enhance the throughput while the job is processing backlog if the source is backlog-aware.

**More Information**
* [FLINK-32514](https://issues.apache.org/jira/browse/FLINK-32514)
* [FLIP-309: Support using larger checkpointing interval when source is processing backlog](https://cwiki.apache.org/confluence/display/FLINK/FLIP-309%3A+Support+using+larger+checkpointing+interval+when+source+is+processing+backlog
```

## CheckpointsCleaner Clean Individual Checkpoint States in Parallel

Now when disposing of no longer needed checkpoints, every state handle/state file will be disposed
in parallel by the ioExecutor, vastly improving the disposing speed of a single checkpoint (for
large checkpoints the disposal time can be improved from 10 minutes to < 1 minute) . The old
behavior can be restored by setting `state.checkpoint.cleaner.parallel-mode` to false.

**More Information**
* [FLINK-33090](https://issues.apache.org/jira/browse/FLINK-33090)

## Trigger Checkpoints through command line client

The command line interface supports triggering a checkpoint manually. Usage:
```shell
./bin/flink checkpoint $JOB_ID [-full]
```
By specifying the '-full' option, a full checkpoint is triggered. Otherwise an incremental
checkpoint is triggered if the job is configured to take incremental ones periodically.

**More Information**
* [FLINK-6755](https://issues.apache.org/jira/browse/FLINK-6755)

# Important Deprecations

> In preparation for the release of Flink 2.0 next year, the community has decided to officially deprecate multiple APIs 

Reviewer 2: ```suggestion
In preparation for the release of Flink 2.0 later this year, the community has decided to officially deprecate multiple APIs 
```

`pipeline.serialization-config`, `pipeline.force-avro`, `pipeline.force-kryo`, and `pipeline.generic-types`. Registration of instance-level serializers is deprecated, using class-level serializers instead.
* We have deprecated all `setXxx` and `getXxx` methods except [`getString(String key, String defaultValue)`](https://github.com/apache/flink/blob/release-1.19/flink-core/src/main/java/org/apache/flink/configuration/Configuration.java#L176) 
and [`setString(String key, String value)`](https://github.com/apache/flink/blob/release-1.19/flink-core/src/main/java/org/apache/flink/configuration/Configuration.java#L220), such as: `setInteger`, `setLong`, `getInteger` and `getLong` etc.
Users and developers are recommend to use get and set methods with `ConfigOption` instead of string as key.
* The non-`ConfigOption` objects in the `StreamExecutionEnvironment`, `CheckpointConfig`, and `ExecutionConfig` and their corresponding getter/setter interfaces are now be deprecated. These objects and methods are planned to be removed in Flink 2.0. The deprecated interfaces include the getter and setter methods of `RestartStrategy`, `CheckpointStorage`, and `StateBackend`.

> * [`org.apache.flink.api.common.functions.RuntimeContext#getExecutionConfig`](https://github.com/apache/flink/blob/release-1.19/flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java#L191) is now officially deprecated and will be dropped in Flink 2.0. Please use [`getGlobalJobParameters()`](https://github.com/apache/flink/blob/release-1.19/flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java#L208) or [`isObjectReuseEnabled()`](https://github.com/apache/flink/blob/release-1.19/flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java#L216).

Reviewer 2: These two are confusing to me.
1. The `ExecutionConfig` is mentioned in both items.
2. Why would we replace `getExecutionConfig)_` with`isObjectReuseEnabled()`?

Reviewer 4: For question 1, there are two separate changes being made:
1. The first change involves deprecating some interfaces in ExecutionConfig that set and get complex Java objects, such as the setRestartStrategy method. 
2. The second change is the deprecation of the getExecutionConfig method in RuntimeContext to prevent the exposure of unnecessary runtime information.

For question 2, I suggest to make the following updates:

`* [`org.apache.flink.api.common.functions.RuntimeContext#getExecutionConfig`](https://github.com/apache/flink/blob/release-1.19/flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java#L191) is now officially deprecated and will be dropped in Flink 2.0. Users are recommend to use the method in RuntimeContext such as [`getGlobalJobParameters()`](https://github.com/apache/flink/blob/release-1.19/flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java#L208), [`createSerializer(TypeInformation<T> typeInformation)`](https://github.com/apache/flink/blob/c6d96b7f7c07faad363779a5175d5772140891a5/flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java#L200) and [`isObjectReuseEnabled()`](https://github.com/apache/flink/blob/release-1.19/flink-core/src/main/java/org/apache/flink/api/common/functions/RuntimeContext.java#L216).`

cc @lincoln-lil

Author: This should be synced with the description on the flip page(https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=278465937)

How about changing it to:
![image](https://github.com/apache/flink-web/assets/3712895/52384271-0065-4fd0-83a4-7897e8a18c75)

Author: @JunRuiLee  I've changed to a list similar to the one in flip(a little clearly for display), what do you think?

Reviewer 4: @lincoln-lil The changes look good to me, thanks for updating!

---
authors:
- LincolnLee:
  name: "Lincoln Lee"
  twitter: lincoln_86xy

date: "2024-03-07T22:00:00Z"
subtitle: ""
title: Announcing the Release of Apache Flink 1.19
aliases:
- /news/2024/03/xx/release-1.19.0.html
---

The Apache Flink PMC is pleased to announce the release of Apache Flink 1.19.0. As usual, we are looking at a packed 
release with a wide variety of improvements and new features. Overall, 166 people contributed to this release completing 
33 FLIPS and 600+ issues. Thank you!

Let's dive into the highlights.

> # Towards a Streaming Lakehouse
> 
> ## Flink SQL Improvements

Reviewer 5: ```suggestion
# Flink SQL Improvements
```

I think this tagline (`Towards a Streaming Lakehouse`) is fine for more promotional writing, but for a set of release notes it 'buries the lede'. This section is about Flink SQL, so we should call it that IMHO. 
I think the fact that it's an H1 followed by H2 and a set of H3s illustrates this point too.

Author: I just followed the previous version of this line. Looking at the main features accomplished in 1.19, it's more of general improvement, so removing it seems reasonable.

> ### Support Setting Parallelism for Table/SQL Sources

Reviewer 5: ```suggestion
## Custom Parallelism for Table/SQL Sources
```

Now in Flink 1.19, you can set a custom parallelism for performance tuning via the `scan.parallelism`

> option. The first available connector is DataGen(Kafka connector is on the way). Here is an example

Reviewer 5: ```suggestion
option. The first available connector is DataGen (Kafka connector is on the way). Here is an example
```

using SQL Client:

```sql
-- set parallelism within the ddl
CREATE TABLE Orders (
    order_number BIGINT,
    price        DECIMAL(32,2),
    buyer        ROW<first_name STRING, last_name STRING>,
    order_time   TIMESTAMP(3)
) WITH (
    'connector' = 'datagen',
    'scan.parallelism' = '4'
);

-- or set parallelism via dynamic table option

> select * from Orders /*+ OPTIONS('scan.parallelism'='4') */;

Reviewer 5: ```suggestion
SELECT * FROM Orders /*+ OPTIONS('scan.parallelism'='4') */;
```
upper case is used elsewhere.

```

**More Information**
* [Documentation](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/sourcessinks/#scan-table-source)
* [FLIP-367: Support Setting Parallelism for Table/SQL Sources](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=263429150)

> ### SQL Gateway Support Configuring Java Options

Reviewer 5: ```suggestion
## Configurable SQL Gateway Java Options
```

A new option `env.java.opts.sql-gateway` for specifying the Java options, so you can fine-tune the
memory settings, garbage collection behavior, and other relevant Java parameters for SQL Gateway.

**More Information**
* [FLINK-33203](https://issues.apache.org/jira/browse/FLINK-33203)

> ### Support Configuring Different State TTLs using SQL Hint

Reviewer 5: ```suggestion
## Configure Different State TTLs using SQL Hint
```

This improvement means that you no longer need to alter your compiled plan to set specific TTLs for
these frequently used operators. With the introduction of `STATE_TTL` hints, you can streamline your workflow and
dynamically adjust the TTL based on your operational requirements.

Here is an example:
```sql
-- set state ttl for join
SELECT /*+ STATE_TTL('Orders'= '1d', 'Customers' = '20d') */ *
FROM Orders LEFT OUTER JOIN Customers
    ON Orders.o_custkey = Customers.c_custkey;

-- set state ttl for aggregation
SELECT /*+ STATE_TTL('o' = '1d') */ o_orderkey, SUM(o_totalprice) AS revenue
FROM Orders AS o
GROUP BY o_orderkey;
```

**More Information**
* [Documentation](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/sql/queries/hints/#state-ttl-hints)
* [FLIP-373: Support Configuring Different State TTLs using SQL Hint](https://cwiki.apache.org/confluence/display/FLINK/FLIP-373%3A+Support+Configuring+Different+State+TTLs+using+SQL+Hint)

> ### Support named parameters for functions and procedures

Reviewer 5: ```suggestion
## Named Parameters for Functions and Procedures
```

```java
public static class NamedArgumentsTableFunction extends TableFunction<Object> {

	@FunctionHint(
			output = @DataTypeHint("STRING"),
			arguments = {
					@ArgumentHint(name = "in1", isOptional = false, type = @DataTypeHint("STRING")),
					@ArgumentHint(name = "in2", isOptional = true, type = @DataTypeHint("STRING")),
					@ArgumentHint(name = "in3", isOptional = true, type = @DataTypeHint("STRING"))})
	public void eval(String arg1, String arg2, String arg3) {
		collect(arg1 + ", " + arg2 + "," + arg3);
	}

}
```
When calling the function in SQL, parameters can be specified by name, for example:
```sql
SELECT * FROM TABLE(myFunction(in1 => 'v1', in3 => 'v3', in2 => 'v2'))
```
Also the optional parameters can be omitted:
```sql
SELECT * FROM TABLE(myFunction(in1 => 'v1'))
```

**More Information**
* [Documentation](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/functions/udfs/#named-parameters)
* [FLIP-387: Support named parameters for functions and call procedures](https://cwiki.apache.org/confluence/display/FLINK/FLIP-387%3A+Support+named+parameters+for+functions+and+call+procedures)

> ### Completes the Unimplemented Window TVF Aggregation Features

Reviewer 5: ```suggestion
## Window TVF Aggregation Features

The implementation of the Window TVF features is now complete.
```

* Supports SESSION Window TVF in Streaming Mode
Now users can use SESSION Window TVF in streaming mode. A simple example is as follows:
```sql
-- session window with partition keys
SELECT * FROM TABLE(
   SESSION(TABLE Bid PARTITION BY item, DESCRIPTOR(bidtime), INTERVAL '5' MINUTES));

-- apply aggregation on the session windowed table with partition keys
SELECT window_start, window_end, item, SUM(price) AS total_price
FROM TABLE(
    SESSION(TABLE Bid PARTITION BY item, DESCRIPTOR(bidtime), INTERVAL '5' MINUTES))
GROUP BY item, window_start, window_end;
```
* Supports Changelog Inputs For Window TVF Aggregation
  Window aggregation operators (generated based on Window TVF Function) can now handle changelog
  streams (e.g., CDC data sources, etc.) without issue. Users are recommended to migrate from legacy
  window aggregation to the new syntax for more complete feature support.

**More Information**
* [Documentation](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/sql/queries/window-tvf/#session)

> ### MiniBatch Optimization for Regular Joins

Reviewer 5: ```suggestion
## Tuning: MiniBatch Optimization for Regular Joins
```

> ##### [FLINK-34219](https://issues.apache.org/jira/browse/FLINK-34219)
> 
> The record amplification is a pain point when performing cascading joins in Flink, now in Flink 1.19,
> the new mini-batch optimization can be used for regular join to reduce intermediate result in such
> cascading join scenarios.
> 
> <div style="text-align: center;">
> <img src="/img/blog/2024-03-xx-release-1.19.0/minibatch_join.png" style="width:90%;margin:15px">
> </div>
> 
> **More Information**

Reviewer 5: ```suggestion
The record amplification is a pain point when performing cascading joins in Flink, now in Flink 1.19,
the new mini-batch optimization can be used for regular join to reduce intermediate result in such
cascading join scenarios.

<div style="text-align: center;">
<img src="/img/blog/2024-03-xx-release-1.19.0/minibatch_join.png" style="width:90%;margin:15px">
</div>

**More Information**
* [FLINK-34219](https://issues.apache.org/jira/browse/FLINK-34219)
```

Reviewer 5: Move the JIRA to the `more information` section

Author: This jira link should be removed

# Key Improvements

## Runtime & Coordination

### Support dynamic source parallelism inference for batch jobs

In Flink 1.19, we have supported dynamic source parallelism inference for batch jobs, which allows
source connectors to dynamically infer the parallelism based on the actual amount of data to consume.
This feature is a significant improvement over previous versions, which only assigned a fixed default
parallelism to source vertices.
Source connectors need to implement the inference interface to enable dynamic parallelism inference.
Currently, the FileSource connector has already been developed with this functionality in place.
Additionally, the configuration `execution.batch.adaptive.auto-parallelism.default-source-parallelism`
will be used as the upper bound of source parallelism inference. And now it will not default to 1.
Instead, if it is not set, the upper bound of allowed parallelism set via
`execution.batch.adaptive.auto-parallelism.max-parallelism` will be used. If that configuration is
also not set, the default parallelism set via `parallelism.default` or `StreamExecutionEnvironment#setParallelism()`
will be used instead.

**More Information**

> * [Documentation](https://nightlies.apache.org/flink/flink-docs-release-1.19/docs/deployment/elastic_scaling/#enable-dynamic-parallelism-inference-support-for-sources).

Reviewer 6: I noticed that some doc links are referred as `https://nightlies.apache.org/flink/flink-docs-release-1.19/`, while some are referred as `https://nightlies.apache.org/flink/flink-docs-master`, can we make them align together?

Author: Of course, I updated to use 1.19 only.

* [FLIP-379: Support dynamic source parallelism inference for batch jobs](https://cwiki.apache.org/confluence/display/FLINK/FLIP-379%3A+Dynamic+source+parallelism+inference+for+batch+jobs)

### Support standard YAML for FLINK configuration

Starting with Flink 1.19, Flink has officially introduced full support for the standard YAML 1.2
syntax. The default configuration file has been changed to `config.yaml` and placed in the `conf/`
directory. Users should directly modify this file to configure Flink.
If users want to use the legacy configuration file `flink-conf.yaml`, users just need to copy this
file into the `conf/` directory. Once the legacy configuration file `flink-conf.yaml` is detected,

> Flink will prioritize using it as the configuration file.And in the upcoming Flink 2.0, the

Reviewer 6: ```suggestion
Flink will prioritize using it as the configuration file. And in the upcoming Flink 2.0, the
```

`flink-conf.yaml` configuration file will no longer work.

**More Information**
* [flink-configuration-file](https://nightlies.apache.org/flink/flink-docs-master/docs/deployment/config/#flink-configuration-file)
* [FLIP-366: Support standard YAML for FLINK configuration](https://cwiki.apache.org/confluence/display/FLINK/FLIP-366%3A+Support+standard+YAML+for+FLINK+configuration?src=contextnavpagetreemode)

> ### Profiling JobManager/TaskManager on Flink Web

Author: @yuchen-ecnu I've added two imgs based on a testing job via latest rc1 package, please help to review the content, thanks!
![image](https://github.com/apache/flink-web/assets/3712895/e1688d9b-fc1b-4544-9e58-ef096dea426b)

### CheckpointsCleaner clean individual checkpoint states in parallel

Now when disposing of no longer needed checkpoints, every state handle/state file will be disposed
in parallel by the ioExecutor, vastly improving the disposing speed of a single checkpoint (for
large checkpoints the disposal time can be improved from 10 minutes to < 1 minute) . The old
behavior can be restored by setting `state.checkpoint.cleaner.parallel-mode` to false.

**More Information**
* [FLINK-33090](https://issues.apache.org/jira/browse/FLINK-33090)

#### Allow triggering Checkpoints through command line client

The command line interface supports triggering a checkpoint manually. Usage:
```shell
./bin/flink checkpoint $JOB_ID [-full]
```
By specifying the '-full' option, a full checkpoint is triggered. Otherwise an incremental
checkpoint is triggered if the job is configured to take incremental ones periodically.

**More Information**
* [FLINK-6755](https://issues.apache.org/jira/browse/FLINK-6755)

# Important Deprecations

In preparation for the release of Flink 2.0 next year, the community has decided to officially deprecate multiple APIs 
that were approaching end of life for a while.

> * Flink's [`Time`](https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/api/common/time/Time.java) is now officially deprecated and will be dropped in Flink 2.0

Reviewer 6: I think we'd better give a full package name, such as `org.apache.flink.api.common.time.Time` to make the description clear.
Moreover, I noticed that the javadoc of `org.apache.flink.api.common.time.Time` is confusing and I had created a ticket [FLINK-34617](https://issues.apache.org/jira/browse/FLINK-34617) to track this problem.

Author: Updated, include the following ones.

Users are encouraged to migrate to the new [`RichFunction#open(OpenContext openContext)`](https://github.com/apache/flink/blob/release-1.19/flink-core/src/main/java/org/apache/flink/api/common/functions/RichFunction.java#L118).
* [`AkkaOptions`](https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/configuration/AkkaOptions.java) is deprecated and replaced with [`RpcOptions`](https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/configuration/RpcOptions.java).

# Upgrade Notes

The Flink community tries to ensure that upgrades are as seamless as possible.
However, certain changes may require users to make adjustments to certain parts
of the program when upgrading to version 1.19. Please refer to the
[release notes](https://nightlies.apache.org/flink/flink-docs-release-1.19/release-notes/flink-1.19/)
for a comprehensive list of adjustments to make and issues to check during the
upgrading process.

# List of Contributors

The Apache Flink community would like to express gratitude to all the
contributors who made this release possible:

> 1996fanrui,

Reviewer 6: I think this id is the same guy as `Rui Fan`.

Adi Polak,
Ahmed Hamdy,
Akira Ajisaka,
Alan Sheinberg,
Aleksandr Pilipenko,
Alex Wu,
Alexander Fedulov,
Archit Goyal,
Asha Boyapati,
Benchao Li,
Bo Cui,
Cheena Budhiraja,
Chesnay Schepler,
Dale Lane,
Danny Cranmer,
David Moravek,
Dawid Wysakowicz,
Deepyaman Datta,
Dian Fu,
Dmitriy Linevich,
Elkhan Dadash,

> Elkhan Dadashov,

Reviewer 5: Are these the same person?

Eric Brzezenski,
Etienne Chauchot,
Fang Yong,
Feng Jiajie,
Feng Jin,
Ferenc Csaky,
Gabor Somogyi,

> GitHub,

Reviewer 6: Maybe this is not a real account.

Gyula Fora,
Hang Ruan,
Hangxiang Yu,
Hanyu Zheng,
Hjw,
Hong Liang Teoh,
Hongshun Wang,
HuangXingBo,
Jack,
Jacky Lau,
James Hughes,
Jane Chan,
Jerome Gagnon,
Jeyhun Karimov,
Jiabao Sun,
JiangXin,
Jiangjie (Becket) Qin,
Jim Hughes,
Jing Ge,
Jinzhong Li,
JunRuiLee,
Laffery,
Leonard Xu,
Lijie Wang,

> Martijn Visser,
> MartijnVisser,

Reviewer 5: ```suggestion
Martijn Visser,
```

Reviewer 5: duplicate

parallelism to source vertices.
Source connectors need to implement the inference interface to enable dynamic parallelism inference.
Currently, the FileSource connector has already been developed with this functionality in place.
Additionally, the configuration `execution.batch.adaptive.auto-parallelism.default-source-parallelism`
will be used as the upper bound of source parallelism inference. And now it will not default to 1.
Instead, if it is not set, the upper bound of allowed parallelism set via
`execution.batch.adaptive.auto-parallelism.max-parallelism` will be used. If that configuration is
also not set, the default parallelism set via `parallelism.default` or `StreamExecutionEnvironment#setParallelism()`
will be used instead.

**More Information**
* [Documentation](https://nightlies.apache.org/flink/flink-docs-release-1.19/docs/deployment/elastic_scaling/#enable-dynamic-parallelism-inference-support-for-sources).
* [FLIP-379: Support dynamic source parallelism inference for batch jobs](https://cwiki.apache.org/confluence/display/FLINK/FLIP-379%3A+Dynamic+source+parallelism+inference+for+batch+jobs)

> ## Standard YAML for FLINK Configuration

Reviewer 7: ```suggestion
## Standard YAML for Flink Configuration
```

---
authors:
- LincolnLee:
  name: "Lincoln Lee"
  twitter: lincoln_86xy

> date: "2024-03-xxT22:00:00Z"

Reviewer 8: minor: add a TODO or FIXME will be better than xx during review phase

## Named Parameters for Functions and Procedures

Named parameters now can be used when calling a function or stored procedure. With named parameters,
users do not need to strictly specify the parameter position, just specify the parameter name and its
corresponding value. At the same time, if non-essential parameters are not specified, they will default to being filled with null.

Here's an example of defining a function with one mandatory parameter and two optional parameters using named parameters:
```java
public static class NamedArgumentsTableFunction extends TableFunction<Object> {

	@FunctionHint(
			output = @DataTypeHint("STRING"),
			arguments = {
					@ArgumentHint(name = "in1", isOptional = false, type = @DataTypeHint("STRING")),
					@ArgumentHint(name = "in2", isOptional = true, type = @DataTypeHint("STRING")),
					@ArgumentHint(name = "in3", isOptional = true, type = @DataTypeHint("STRING"))})
	public void eval(String arg1, String arg2, String arg3) {
		collect(arg1 + ", " + arg2 + "," + arg3);
	}

}
```
When calling the function in SQL, parameters can be specified by name, for example:
```sql

> SELECT * FROM TABLE(myFunction(in1 => 'v1', in3 => 'v3', in2 => 'v2'))

Reviewer 6: Another small suggestion, we'd better add a semicolon in the end of SQL statement.
`SELECT * FROM TABLE(myFunction(in1 => 'v1', in3 => 'v3', in2 => 'v2'));`

```
Also the optional parameters can be omitted:
```sql

> SELECT * FROM TABLE(myFunction(in1 => 'v1'))

Reviewer 6: Same here: `SELECT * FROM TABLE(myFunction(in1 => 'v1'));`

## CheckpointsCleaner Clean Individual Checkpoint States in Parallel

Now when disposing of no longer needed checkpoints, every state handle/state file will be disposed
in parallel by the ioExecutor, vastly improving the disposing speed of a single checkpoint (for
large checkpoints the disposal time can be improved from 10 minutes to < 1 minute) . The old
behavior can be restored by setting `state.checkpoint.cleaner.parallel-mode` to false.

**More Information**
* [FLINK-33090](https://issues.apache.org/jira/browse/FLINK-33090)

## Trigger Checkpoints through Command Line Client

The command line interface supports triggering a checkpoint manually. Usage:
```shell
./bin/flink checkpoint $JOB_ID [-full]
```
By specifying the '-full' option, a full checkpoint is triggered. Otherwise an incremental
checkpoint is triggered if the job is configured to take incremental ones periodically.

**More Information**
* [FLINK-6755](https://issues.apache.org/jira/browse/FLINK-6755)

# Connector API Improvements

> ## New Interfaces to SinkV2 That Are Consistent with SourceV2

Reviewer 9: What about `Source` instead of `SourceV2`? We never introduced a v2 API for source

Author: @PatrickRen Thanks for the correction! I directly use the title of FLINK-33973, but it's true that we don't have a class like `Sourcev2` in the source code, IIUC, "v2 " source should actually refer to FLIP-27 source.

Author: updated
